dod-wrapper/src/main/java/com/revolut/rxdata/dod_wrapper/DataObservableDelegateWrapper.kt

```diff
@@ -0,0 +1,117 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+class DataObservableDelegateWrapper<Params : Any, Domain : Any>(
+    fromNetwork: suspend DataObservableDelegateWrapper<Params, Domain>.(params: Params) -> Domain,
+    fromMemory: (params: Params) -> Domain?,
+    toMemory: (params: Params, Domain) -> Unit,
+    fromStorage: suspend (params: Params) -> Domain?,
+    toStorage: suspend (params: Params, Domain) -> Unit,
+    onRemove: suspend (params: Params) -> Unit = { _ -> Unit },
+) {
+    private val inner = DataObservableDelegate<Params, Domain>(
+        fromNetwork = { params ->
+            rxSingle(DataObservableDelegateWrapperDispatchers.Unconfined) {
+                fromNetwork(params)
+            }
+        },
+        fromMemory = { params ->
+            fromMemory(params)
+        },
+        toMemory = { params, domain ->
+            toMemory(params, domain)
+        },
+        fromStorage = { params ->
+            runBlocking {
+                fromStorage(params)
+            }
+        },
+        toStorage = { params, domain ->
+            runBlocking {
+                toStorage(params, domain)
+            }
+        },
+        onRemove = { params ->
+            runBlocking {
+                onRemove(params)
+            }
+        }
+    )
+
+    fun observe(params: Params, forceReload: Boolean = true): Flow<Data<Domain>> =
+        inner.observe(
+            params = params,
+            forceReload = forceReload,
+        ).asFlow()
+
+    /**
+     * Replaces the data in both caches (Memory, Persistent storage)
+     * and emits an update.
+     */
+    suspend fun updateAll(params: Params, domain: Domain) =
+        withContext(DataObservableDelegateWrapperDispatchers.IO) {
+            inner.updateAll(params = params, domain = domain)
+        }
+
+    /**
+     * Replaces the data and emits an update in persistent storage cache.
+     *
+     * /!\ Memory cache won't be dropped or replaced /!\
+     */
+    suspend fun updateStorage(params: Params, domain: Domain) =
+        withContext(DataObservableDelegateWrapperDispatchers.IO) {
+            inner.updateStorage(params = params, domain = domain)
+        }
+
+    /**
+     * Replaces the data and emits an update in memory cache.
+     */
+    fun updateMemory(params: Params, domain: Domain) =
+        inner.updateMemory(params = params, domain = domain)
+
+    /**
+     * Subscribers observing this DOD will be notified with
+     * Data(fromMemory(params), loading = false, error = null).
+     * @param where must return true if subscriber requires notification.
+     */
+    fun notifyFromMemory(
+        error: Throwable? = null,
+        loading: Boolean = false,
+        where: (Params) -> Boolean
+    ) = inner.notifyFromMemory(error = error, loading = loading, where = where)
+
+    suspend fun remove(params: Params) =
```

Comment:
I don't think most of these one need to be suspendable, their synchronous and use serialised publish processor underhood, and proxies the call to lambdas (like onRemove) so they never rally had any threading requirements on where to call them from. Only exception is reload which needs to await() a Completable.

Reply:
This was raised by @PavloRatushnyi and I suppose he has a valid point: these methods update storage under the hood, so they are expected to perform I/O. So in DOD they have an implicit threading requirement: to be called from background thread, which is error-prone.

In the coroutines world such functionality is typically exposed as ["main-safe" suspending functions](https://developer.android.com/kotlin/coroutines#use-coroutines-for-main-safety), to prevent callers from accidentally blocking UI thread.